README.md
# AnimatedReadMoreText

A Flutter package that provides a customizable widget for displaying text with an optional "Read more/Read less" button, enhancing the user experience for text-heavy content.

## Overview

The **AnimatedReadMoreText** widget is a Flutter package that provides a user-friendly and visually appealing way to present lengthy text content. It dynamically adapts text length based on a predefined maximum line count, ensuring optimal readability on various screen sizes. The widget's key feature is its ability to seamlessly expand or collapse text based on user interaction. This functionality enhances readability by allowing users to focus on specific sections of text without being overwhelmed by lengthy passages. Additionally, the widget's subtle yet impactful animations add a touch of elegance to the user interface, further enhancing the overall reading experience.


![AnimatedReadMoreText](https://github.com/monster555/animated_read_more_text/assets/32662133/9768e6a7-ed68-4010-8c79-a2c6a4026862)



## Features

- Lightweight and easy to use
- Displays text with a specified maximum number of lines.
- Set custom labels for the "Read more" and "Read less" buttons to match your app's branding and enhance the user experience.
- Customizable appearance and behavior.
- Supports optional styling for the text and button.
- Smooth and customizable animations for text expansion and collapse.

## Usage

To use this package, include it in your `pubspec.yaml` file:

```yaml
dependencies:
  animated_read_more_text: ^0.0.6
```
Import the package in your Dart file:

```dart
import 'package:animated_read_more_text/animated_read_more_text.dart';
```

Use the AnimatedReadMoreText widget in your Flutter application:
```dart
AnimatedReadMoreText(
  yourLongText,
  maxLines: 3,
  // Set a custom text for the expand button. Defaults to Read more
  readMoreText: 'Expand',
  // Set a custom text for the collapse button. Defaults to Read less
  readLessText: 'Collapse',
  // Set a custom text style for the main block of text
  textStyle: TextStyle(
    fontSize: 16,
    color: Colors.red,
  ),
  // Set a custom text style for the expand/collapse button
  buttonTextStyle: TextStyle(
    fontSize: 16,
    fontWeight: FontWeight.bold,
  ),
)
```

Set custom animation curve and duration
```dart
AnimatedReadMoreText(
  yourLongText,
  // Set the curve of the animation
  animationCurve: Curves.easeInOut,
  // Set the duration of the animation
  animationDuration: Duration(milliseconds: 500),
)
```
## Contributions
Contributions and bug reports are welcome! Please feel free to create an issue or submit a pull request.

## License
This package is licensed under the MIT License - see the LICENSE file for details.


CHANGELOG.md
## 0.0.4

* Updated the README with the package version.

## 0.0.3

* Updated the README with an animated GIF to demonstrate package functionality.

## 0.0.2

* Added supported platforms

## 0.0.1

* Initial release


test/test_helpers.dart
import 'package:animated_read_more_text/animated_read_more_text.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

/// This function returns a Finder that matches [RichText] widgets containing the specified text.
///
/// The function uses a widget predicate to filter widgets. The predicate takes a [Widget] and checks if it's a [RichText] widget. If it is, the function checks if the plain text of the [RichText] widget contains the specified text. If the text is found, the function returns true, indicating that the widget matches the predicate.
///
/// This function is useful for finding [RichText] widgets in widget tests. For example, it can be used to find a [RichText] widget representing a button with a specific label.
///
/// - [text] The text to search for in the [RichText] widgets.
/// - Return a Finder that matches [RichText] widgets containing the specified text.
Finder findRichTextWithText(String text) => find.byWidgetPredicate(
      (Widget widget) {
        // Check if the widget is a RichText widget
        if (widget is RichText) {
          // If it is, check if the plain text of the RichText widget contains the specified text
          return widget.text.toPlainText().contains(text);
        }
        // If the widget is not a RichText widget, it doesn't match the predicate
        return false;
      },
    );

/// This function builds and renders an [AnimatedReadMoreText] widget in a test environment.
///
/// The function takes a [WidgetTester] and a string of text to display in the widget. It also takes optional parameters for the [maxLines], [expandOnTextTap], and [allowCollapse] properties of the [AnimatedReadMoreText] widget. The function uses these parameters to create an [AnimatedReadMoreText] widget, which it then renders using the [pumpWidget] method of the [WidgetTester].
///
/// This function is useful for reducing code duplication in widget tests. Instead of creating and rendering an [AnimatedReadMoreText] widget in each individual test, you can call this function with the desired parameters.
///
/// - [tester] The [WidgetTester] to use for rendering the widget.
/// - [text] The text to display in the [AnimatedReadMoreText] widget.
/// - [maxLines] The maximum number of lines to display in the [AnimatedReadMoreText] widget.
/// - [expandOnTextTap] Whether the [AnimatedReadMoreText] widget should expand when the text is tapped.
/// - [allowCollapse] Whether the [AnimatedReadMoreText] widget should allow collapsing.
/// - Return a [Future] that completes when the widget has been rendered.
Future<void> pumpAnimatedReadMoreText(
  WidgetTester tester,
  String text, {
  int maxLines = 3,
  bool expandOnTextTap = true,
  bool allowCollapse = true,
}) async {
  await tester.pumpWidget(MaterialApp(
    home: Scaffold(
      body: AnimatedReadMoreText(
        text,
        maxLines: maxLines,
        expandOnTextTap: expandOnTextTap,
      ),
    ),
  ));
}


test/animated_read_more_test.dart
import 'package:animated_read_more_text/animated_read_more_text.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'test_helpers.dart';

void main() {
  const String longText =
      'Underneath the vast canopy of the ancient forest, a gentle brook meandered its way through the undergrowth. The sunlight filtered through the leaves, casting dappled shadows on the forest floor. Birds chirped merrily from the treetops, their songs echoing through the stillness. A deer grazed peacefully by the water’s edge, its ears twitching at the slightest sound. This was a world untouched by time, a symphony of nature playing its eternal melody.';
  const String shortText =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit.';

  group('AnimatedReadMoreText widget', () {
    testWidgets('Test AnimatedReadMoreText widget',
        (WidgetTester tester) async {
      // Build your widget
      await pumpAnimatedReadMoreText(tester, longText);

      // Check if the widget is in the tree
      expect(find.byType(AnimatedReadMoreText), findsOneWidget);

      // Check if the LayoutBuilder is in the tree
      expect(find.byType(LayoutBuilder), findsOneWidget);

      // Check if the GestureDetector is in the tree
      expect(find.byType(GestureDetector), findsOneWidget);

      // Check if the AnimatedCrossFade is in the tree
      expect(find.byType(AnimatedCrossFade), findsOneWidget);
    });
  });

  group('maxLines property', () {
    testWidgets('Test AnimatedReadMoreText widget with short text',
        (WidgetTester tester) async {
      // Build your widget
      await pumpAnimatedReadMoreText(tester, shortText);

      // Check if the widget is in the tree
      expect(find.byType(AnimatedReadMoreText), findsOneWidget);

      // Check if the "Show more" button is not in the tree
      expect(findRichTextWithText('Show more'), findsNothing);

      // Check if the "Show less" button is not in the tree
      expect(findRichTextWithText('Show less'), findsNothing);
    });
  });

  group('expandOnTextTap property', () {
    // Test the expandOnTextTap property set to true
    testWidgets('expandOnTextTap property set to true',
        (WidgetTester tester) async {
      await pumpAnimatedReadMoreText(tester, longText, expandOnTextTap: true);

      // Check the crossFadeState of the AnimatedCrossFade widget
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showFirst);

      // Get the bounding box of the text block
      final textBlockRect = tester.getRect(find.byType(AnimatedReadMoreText));

      // Calculate the top-left point of the text block
      final topLeft = textBlockRect.topLeft;

      // Tap the top-left point of the text block to expand it
      await tester.tapAt(topLeft);
      await tester.pumpAndSettle();

      // Check the crossFadeState of the AnimatedCrossFade widget
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showSecond);

      // Tap the top-left point of the text block to collapse it
      await tester.tapAt(topLeft);
      await tester.pumpAndSettle();

      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showFirst);
    });

    // Test the expandOnTextTap property set to false
    testWidgets('expandOnTextTap property set to false',
        (WidgetTester tester) async {
      await pumpAnimatedReadMoreText(tester, longText, expandOnTextTap: false);

      // Check the crossFadeState of the AnimatedCrossFade widget
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showFirst);

      // Get the bounding box of the text block
      final textBlockRect = tester.getRect(find.byType(AnimatedReadMoreText));

      // Calculate the top-left point of the text block
      final topLeft = textBlockRect.topLeft;

      // Tap the top-left point of the text block to expand it
      // It should not expand because expandOnTextTap is set to false
      await tester.tapAt(topLeft);
      await tester.pumpAndSettle();

      // Check the crossFadeState of the AnimatedCrossFade widget
      // Verify that the text block is still collapsed.
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showFirst);
    });
  });

  group('expand and collapse of AnimatedReadMoreText', () {
    testWidgets(
        'expand and collapse of AnimatedReadMoreText widget b ytapping the Read more and Read less buttons',
        (WidgetTester tester) async {
      // Build your widget
      await pumpAnimatedReadMoreText(tester, longText);

      // Check if the widget is in the tree
      expect(find.byType(AnimatedReadMoreText), findsOneWidget);

      // Check if the "Show more" button is in the tree
      expect(findRichTextWithText('Show more'), findsOneWidget);

      // Check the crossFadeState of the AnimatedCrossFade widget
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showFirst);

      // Tap the "Show more" button
      await tester.tap(findRichTextWithText('Show more'));
      await tester.pumpAndSettle();

      // Check if the "Show less" button is in the tree
      expect(findRichTextWithText('Show less'), findsOneWidget);

      // Check the crossFadeState of the AnimatedCrossFade widget
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showSecond);

      // Tap the "Show less" button
      await tester.tap(findRichTextWithText('Show less'));
      await tester.pumpAndSettle();

      // Check if the "Show more" button is in the tree again
      expect(findRichTextWithText('Show more'), findsOneWidget);

      // Check the crossFadeState of the AnimatedCrossFade widget
      expect(
          tester
              .widget<AnimatedCrossFade>(find.byType(AnimatedCrossFade))
              .crossFadeState,
          CrossFadeState.showFirst);
    });

    // Test the allowCollapse property
    testWidgets('allowCollapse property', (WidgetTester tester) async {
      await pumpAnimatedReadMoreText(tester, longText, allowCollapse: false);

      // Check if the "Show less" button is not in the tree
      expect(find.text('Show less'), findsNothing);

      // Check that the widget does not collapse the text once it has been expanded
      // Tap the "Show more" button
      await tester.tap(findRichTextWithText('Show more'));
      await tester.pumpAndSettle();

      // Check if the "Show less" button is not in the tree
      expect(find.text('Show less'), findsNothing);
    });
  });
}


lib/animated_read_more_text.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

/// A Flutter widget for displaying text with an optional "Read more/Read less" button.
///
/// The `AnimatedReadMoreText` widget allows you to present text with a maximum
/// number of lines and provides a button to expand or collapse the content for
/// improved readability.
///
/// ## Usage
///
/// To use this widget, provide the required `text` parameter with the content you
/// want to display. Customize the appearance and behavior by adjusting various
/// optional parameters such as `maxLines`, `readMoreText`, `readLessText`,
/// `textStyle`, and more.
class AnimatedReadMoreText extends StatefulWidget {
  /// Creates an AnimatedReadMoreText widget.
  ///
  /// This widget is designed for displaying text with optional "Read more/Read less"
  /// functionality. It provides a convenient way to present content with a maximum
  /// number of visible lines and offers a button to expand or collapse the text for
  /// improved user interaction.
  ///
  /// ## Example
  ///
  /// ```dart
  /// AnimatedReadMoreText(
  ///   'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
  ///   maxLines: 3,
  ///   readMoreText: 'Show more', // default: 'Read more'
  ///   readLessText: 'Show less', // default: 'Read less'
  /// )
  /// ```
  const AnimatedReadMoreText(
    this.text, {
    super.key,
    this.maxLines = 3,
    this.readMoreText = 'Read more',
    this.readLessText = 'Read less',
    this.textStyle,
    this.buttonTextStyle,
    this.isSuffixButton = true,
    this.allowCollapse = true,
    this.expandOnTextTap = true,
    this.animationCurve = Curves.easeOutBack,
    this.animationDuration = const Duration(milliseconds: 300),
  })  : assert(maxLines > 1, 'maxLines must be greater than 1'),
        assert(text.length > 0, 'text cannot be empty'),
        assert(readMoreText.length > 0, 'readMoreText cannot be empty'),
        assert(readLessText.length > 0, 'readLessText cannot be empty');

  /// The text to be displayed in the widget.
  final String text;

  /// The maximum number of lines to display before collapsing the text.
  /// Must be greater than 1.
  final int maxLines;

  /// The text to display for the "Read more" button.
  /// Must not be an empty string.
  final String readMoreText;

  /// The text to display for the "Read less" button.
  /// Must not be an empty string.
  final String readLessText;

  /// The style to apply to the "Read more/Read less" button.
  final TextStyle? buttonTextStyle;

  /// The style to apply to the displayed text.
  final TextStyle? textStyle;

  /// Determines if the "Read more/Read less" button is displayed as a suffix.
  final bool isSuffixButton;

  /// Controls whether the text can be collapsed.
  final bool allowCollapse;

  /// Enables text expansion when tapped.
  final bool expandOnTextTap;

  /// The animation curve for expanding and collapsing the text.
  final Curve animationCurve;

  /// The duration of the animation.
  final Duration animationDuration;

  @override
  State<AnimatedReadMoreText> createState() => _AnimatedReadMoreTextState();
}

class _AnimatedReadMoreTextState extends State<AnimatedReadMoreText> {
  /// Whether the text is expanded.
  bool _isExpanded = false;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        assert(constraints.hasBoundedWidth,
            'The parent widget must provide bounded width constraints');

        // Obtain the text style that will be used for rendering. If the widget doesn't provide a specific style,
        // use the default style of the current text context.
        final style = widget.textStyle ?? DefaultTextStyle.of(context).style;
        final fontSize =
            style.fontSize ?? DefaultTextStyle.of(context).style.fontSize!;

        // Form a TextSpan object representing the text to be displayed, along with its associated style.
        final textSpan = TextSpan(text: widget.text, style: style);

        // Initialize a TextPainter to measure and layout the text. Set constraints such as the maximum number
        // of lines and text direction.
        final textPainter = TextPainter(
          text: textSpan,
          maxLines: widget.maxLines,
          textDirection: TextDirection.ltr,
        )..layout(maxWidth: constraints.maxWidth);

        final bool showButton = textPainter.didExceedMaxLines &&
            (widget.allowCollapse || !_isExpanded);

        // Find the position of the last character within the available width and considering the maximum number of lines.
        final endPosition = textPainter.getPositionForOffset(Offset(
          constraints.maxWidth,
          // Determine whether the text should be truncated based on the current expansion state.
          // If not expanded, create a substring of the text up to the calculated end position;
          // otherwise, use the full text.
          fontSize * widget.maxLines,
        ));

        // Determine the truncated text based on the current expansion state.
        final truncatedText = !_isExpanded
            ? widget.text.substring(0, endPosition.offset)
            : widget.text;

        return GestureDetector(
          onTap: () {
            if (widget.expandOnTextTap &&
                (widget.allowCollapse || !_isExpanded)) {
              _toggleExpanded();
            }
          },
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              AnimatedCrossFade(
                firstChild: _ExpandableTextSpan(
                  showButton: showButton,
                  text: truncatedText,
                  style: widget.textStyle,
                  buttonTextStyle: widget.buttonTextStyle,
                  readMoreLessText: widget.readMoreText,
                  isSuffixButton: true,
                  // isSuffixButton: widget.isSuffixButton,
                  onPressed: _toggleExpanded,
                ),
                secondChild: _ExpandableTextSpan(
                  showButton: showButton,
                  text: widget.text,
                  style: widget.textStyle,
                  buttonTextStyle: widget.buttonTextStyle,
                  readMoreLessText: widget.readLessText,
                  isSuffixButton: true,
                  // isSuffixButton: widget.isSuffixButton && widget.allowCollapse,
                  onPressed: _toggleExpanded,
                ),
                crossFadeState: _isExpanded
                    ? CrossFadeState.showSecond
                    : CrossFadeState.showFirst,
                duration: widget.animationDuration,
                sizeCurve: widget.animationCurve,
              ),
              if (!widget.isSuffixButton &&
                  (widget.allowCollapse || !_isExpanded) &&
                  showButton)
                ReadMoreButton(
                  isExpanded: _isExpanded,
                  readMoreText: widget.readMoreText,
                  readLessText: widget.readLessText,
                  onPressed: _toggleExpanded,
                ),
            ],
          ),
        );
      },
    );
  }

  /// Toggles the expansion state of the widget.
  ///
  /// This method flips the value of [_isExpanded] and triggers a widget rebuild
  /// to reflect the updated state.
  void _toggleExpanded() => setState(() => _isExpanded = !_isExpanded);
}

/// A widget that creates an expandable text span with an optional "Read more/Read less" button.
///
/// The `ExpandableTextSpan` widget allows you to present a portion of text with an optional
/// button to expand or collapse the content for improved readability.
class _ExpandableTextSpan extends StatelessWidget {
  /// Creates an ExpandableTextSpan widget.
  ///
  /// To use this widget, provide the required parameters such as `text`,
  /// `readMoreLessText`, `isSuffixButton`, and `onPressed`. Customize the
  /// appearance and behavior by adjusting optional parameters like `style` and
  /// `buttonTextStyle`.
  const _ExpandableTextSpan({
    required this.showButton,
    required this.text,
    this.style,
    this.buttonTextStyle,
    required this.readMoreLessText,
    required this.isSuffixButton,
    required this.onPressed,
  });

  /// A boolean indicating whether the button should be displayed.
  final bool showButton;

  /// The text to be displayed in the widget.
  final String text;

  /// The style to apply to the displayed text.
  final TextStyle? style;

  /// The style to apply to the "Read more/Read less" button.
  final TextStyle? buttonTextStyle;

  /// The text to display for the "Read more/Read less" button.
  final String readMoreLessText;

  /// Determines if the button is displayed as a suffix.
  final bool isSuffixButton;

  /// The callback function executed when the button is tapped.
  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return Text.rich(
      TextSpan(
        text: text,
        style: style,
        children: [
          if (isSuffixButton && showButton)
            _buildExpandCollapseTextSpan(' $readMoreLessText'),
        ],
      ),
    );
  }

  /// Builds a [TextSpan] for the "Read more/Read less" button.
  ///
  /// This method constructs a [TextSpan] with the provided `text`, `buttonTextStyle`,
  /// and an [onPressed] action using a [TapGestureRecognizer].
  TextSpan _buildExpandCollapseTextSpan(String text) {
    return TextSpan(
      text: text,
      style: buttonTextStyle ??
          const TextStyle(
            fontWeight: FontWeight.bold,
          ),
      recognizer: TapGestureRecognizer()..onTap = onPressed,
    );
  }
}

/// A widget that represents a "Read more/Read less" button.
///
/// The `ReadMoreButton` widget provides a button that changes its text based
/// on whether the associated content is expanded or collapsed.
class ReadMoreButton extends StatelessWidget {
  /// Creates a ReadMoreButton widget.
  ///
  /// To use this widget, provide the required parameters such as `isExpanded`,
  /// `readMoreText`, `readLessText`, and `onPressed`.
  const ReadMoreButton({
    super.key,
    required this.isExpanded,
    required this.readMoreText,
    required this.readLessText,
    required this.onPressed,
  });

  /// A boolean indicating whether the associated content is expanded.
  final bool isExpanded;

  /// The text to display when the content is collapsed.
  final String readMoreText;

  /// The text to display when the content is expanded.
  final String readLessText;

  /// The callback function executed when the button is tapped.
  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Padding(
        padding: const EdgeInsets.only(top: 16.0, bottom: 8.0),
        child: TextButton(
          onPressed: onPressed,
          child: Text(
            isExpanded ? readLessText : readMoreText,
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:animated_read_more_text/animated_read_more_text.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MainApp());

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('AnimatedReadMoreText Example'),
        ),
        body: const Padding(
          padding: EdgeInsets.all(8.0),
          child: Column(
            children: <Widget>[
              // Example 1: Basic usage
              AnimatedReadMoreText(
                'This is a short text that will be displayed in its entirety because it does not exceed the maximum number of lines.',
              ),

              SizedBox(height: 16),

              // Example 2: Long text with custom styles, custom "Read more" and "Read less" buttons text, and no expand on text tap
              AnimatedReadMoreText(
                'This text exceeds the maximum number of lines. The text will be truncated, and an "Expand" button will appear, replacing the default "Read more" button. The text and "Expand" button are styled with a custom font size and color. The text is blue with a font size of 16.0, and the "Expand" button is red with a font size of 14.0.',
                maxLines: 3,
                readMoreText: 'Expand',
                readLessText: 'Collapse',
                textStyle: TextStyle(fontSize: 16.0, color: Colors.blue),
                buttonTextStyle: TextStyle(fontSize: 14.0, color: Colors.red),
                expandOnTextTap: false,
              ),

              SizedBox(height: 16),

              // Example 3: Expand on text tap
              AnimatedReadMoreText(
                'This text block is designed to be user-friendly. By tapping anywhere on the text, not just the ‘Read more’ button, you can expand the content. This feature enhances the user experience by making the entire text block interactive. It provides an intuitive way to access additional content. So, tap on the text and discover more!',
                expandOnTextTap: true,
              ),

              SizedBox(height: 16),

              // Example 4: Non-collapsible text
              AnimatedReadMoreText(
                'This text can be expanded, but not collapsed. Once the "Read more" button is tapped, the text will be displayed in its entirety and cannot be truncated again. This is useful if you want to ensure that the user sees the full text once they have chosen to expand it.',
                allowCollapse: false,
              ),
            ],
          ),
        ),
      ),
    );
  }
}
